packages/@sanity/presence-store/src/PresenceStore.js

```diff
@@ -0,0 +1,152 @@
+
+function hashKeyForState(identity, session) {
+  return `${identity}__${session}`
+}
+
+export default class PresenceStore {
+  constructor(connection, channel) {
+    this.connection = connection
+    this.unsubscribeListener = this.connection.listen().subscribe(this.handleMessage).unsubscribe
+    this.myState = {}
+    this.states = {}
+    this.timestamps = {}
+    this.resendReportTimer = null
+    this.changeReportDebounceTimer = null
+    this.performPurgeTimer = setInterval(this.performPurge, 2000)
+    this.sessionId = UUID()
+    this.requestRollCall()
+
+    // List of functions that will be called with presence change notifications
+    this.subscriberCallbacks = []
+  }
+
+  // Close Presence Store, should be called when window closes
+  close() {
+    this.connection.sendBeacon({type: 'disconnect', session: this.sessionId})
+    this.unsubscribeListener()
+    clearTimeout(this.resendReportTimer)
+    clearInterval(this.performPurgeTimer)
+  }
+
+  // The observable that all subscribers will use to track state changes
+  presence = new Observable(observer => {
+    // Send initial state
+    observer.next(this.getStateReport())
+
+    // Set up subscription to track subsequent changes
+    const reporter = value => {
+      observer.next(value)
+    }
+    this.subscriberCallbacks.push(reporter)
+
+    // Create and return unsubscriber
+    const unsubscribe = () => {
+      this.subscriberCallbacks = this.subscriberCallbacks.filter(cb => cb != reporter)
+    }
+    return unsubscribe
+  })
+
+  // Call this to report this clients state
+  reportMyState(state) {
+    this.myState = state
+    this.sendMyState()
+  }
+
+  // Updates state for a client based on an incoming state message
+  updateClientState(msg) {
+    // Construct next state object
+    const state = Object.assign({
+      identity: msg.i
+    }, msg.m)
+    delete state.type
+
+    // The cache key for this state
+    const key = hashKeyForState(msg.i, msg.m.session)
+
+    // Only update state if state is actually changed
+    if (!deepEqual(state, this.states[key])) {
+      this.states[key] = state
+      this.handleRemoteChange()
+    }
+    // Update timestamp for this cache key
+    this.timestamps[key] = new Date()
+  }
+
+  handleMessage = msg => {
+    // Ignore messages from ourselves
+    if (msg.m.session == this.sessionId) {
+      return
+    }
+    switch (msg.m.type) {
+      case 'state': {
+        this.updateClientState(msg)
+        return
+      }
+      case 'rollCall':
+        this.sendMyState()
+        break
+      case 'disconnect':
+        delete this.states[hashKeyForState(msg.i, msg.m.session)]
+        this.handleRemoteChange()
+        break
+      default:
+    }
+  }
+
+  // Purge state objects for clients that have not reported for STATE_TIMEOUT_INTERVAL millis
+  performPurge = () => {
+    const now = new Date()
+    Object.keys(this.states).forEach(key => {
+      const age = now - this.timestamps[key].timestamp
```

Comment:
Don't think `timestamp` is a standardized property, but as pointed out earlier, a `Date.now()` would skip the need for this